lib/rules/no-this-in-template-only-components.js

```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const Rule = require('./base');
+
+function message(original, fixed) {
+  return (
+    `Usage of 'this' in path '${original}' is not allowed in a template-only component. ` +
+    `Use '${fixed}' if it is a named argument or create a component.js for this component.`
+  );
+}
+
+function hasContext(hbsFilePath) {
+  const paths = hbsFilePath.split(path.sep);
+  if (paths[0] === 'app' || paths[0] === 'addon') {
+    if (paths[1] === 'templates') {
+      if (paths[2] === 'components') {
+        // classic structure: /app/templates/components/foo.hbs => /app/components/foo.js
+        const moduleName = path.basename(hbsFilePath, '.hbs');
+        const jsFilePath = path.join(
+          paths[0],
+          'components',
+          ...paths.slice(3, -1),
+          `${moduleName}.js`
+        );
+        return fs.existsSync(jsFilePath);
+      } else {
+        // route template always has a context
+        return true;
+      }
+    } else if (paths[1] === 'components') {
+      // co-located structure: /app/components/foo.hbs => /app/components/foo.js
+      const moduleName = path.basename(hbsFilePath, '.hbs');
+      const jsFilePath = path.join(path.dirname(hbsFilePath), `${moduleName}.js`);
+      return fs.existsSync(jsFilePath);
+    }
+  }
+  // what other types of files?
+  return false;
+}
+
+function isFixable(original) {
+  if (original === 'this.elementId') {
+    // elementId usage in a template-only component can't be auto-fixed
+    return false;
```

Comment:
Can you extract this check to check for any of the "built in" properties? 

I think we can figure this out by looking through 

https://github.com/emberjs/ember.js/blob/2bb644532d407d4b4700a81f510c5aed4caf5d84/packages/%40ember/-internals/glimmer/lib/component.ts#L647-L652

And figuring out all of the properties defined on those mixins. I _think_ this is the full list:

```
childViews
concatenatedProperties
attributeBindings
element
elementId
tagName
classNames
classNameBindings
target
actionContextObject
action
actionContext
```

Reply:
I added those plus `parentView`